csp/baselib.py

```diff
@@ -759,32 +880,101 @@ def comp(x: ts["T"], y: ts["U"]):
+    def float_type(x: ts[float]) -> ts[float_out_type]:
+        if csp.valid(x):
+            return op_lambda(x)
 
-eq = define_op("eq", lambda x, y: x == y)
-ne = define_op("ne", lambda x, y: x != y)
-gt = define_op("gt", lambda x, y: x > y)
-lt = define_op("lt", lambda x, y: x < y)
-ge = define_op("ge", lambda x, y: x >= y)
-le = define_op("le", lambda x, y: x <= y)
+    @_node_internal_use(cppimpl=MATH_COMP_OPS_CPP.get((name, "int"), None), name=name)
+    def int_type(x: ts[int]) -> ts[int_out_type]:
+        if csp.valid(x):
+            return op_lambda(x)
 
-# Other math ops
+    numpy_func = MATH_COMP_OPS_CPP.get((name, "np"), op_lambda)
+
+    @_node_internal_use(name=name)
+    def numpy_type(x: ts["T"]) -> ts[np.ndarray]:
+        if csp.valid(x):
+            return numpy_func(x)
 
+    @_node_internal_use(name=name)
+    def generic_type(x: ts["T"]) -> ts[generic_out_type]:
+        if csp.valid(x):
+            return op_lambda(x)
+
+    def comp(x: ts["T"]):
+        if typing.get_origin(x.tstype.typ) in [Numpy1DArray, NumpyNDArray]:
+            return numpy_type(x)
+        elif x.tstype.typ is float:
+            return float_type(x)
+        elif x.tstype.typ is int:
+            return int_type(x)
+        return generic_type(x)
 
-@node
-def floordiv(x: ts["T"], y: ts["T"]) -> ts["T"]:
-    if csp.ticked(x, y) and csp.valid(x, y):
-        return x // y
+    comp.__name__ = name
+    return comp
+
+
+# Math operators
+add = define_binary_op("add", lambda x, y: x + y)
+sub = define_binary_op("sub", lambda x, y: x - y)
+multiply = define_binary_op("multiply", lambda x, y: x * y)
+divide = define_binary_op("divide", lambda x, y: x / y)
+pow = define_binary_op("pow", lambda x, y: x**y)
+min = define_binary_op("min", lambda x, y: x if x < y else y)
+max = define_binary_op("max", lambda x, y: x if x > y else y)
+floordiv = define_binary_op("floordiv", lambda x, y: x // y)
+mod = define_binary_op("mod", lambda x, y: x % y)
+neg = define_unary_op("mod", lambda x: -x)
+
+# Other math ops
+_python_abs = abs
+abs = define_unary_op("abs", lambda x: _python_abs(x))
+ln = define_unary_op("ln", lambda x: math.log(x))
+log2 = define_unary_op("log2", lambda x: math.log2(x))
+log10 = define_unary_op("log10", lambda x: math.log10(x))
+exp = define_unary_op("exp", lambda x: math.exp(x))
+exp2 = define_unary_op("exp2", lambda x: math.exp2(x))
+sqrt = define_unary_op("sqrt", lambda x: math.sqrt(x))
+erf = define_unary_op("erf", lambda x: math.erf(x))
+sin = define_unary_op("sin", lambda x: math.sin(x))
+cos = define_unary_op("cos", lambda x: math.cos(x))
+tan = define_unary_op("tan", lambda x: math.tan(x))
+arcsin = define_unary_op("arcsin", lambda x: math.asin(x))
+arccos = define_unary_op("arccos", lambda x: math.acos(x))
+arctan = define_unary_op("arctan", lambda x: math.atan(x))
+sinh = define_unary_op("sinh", lambda x: math.sinh(x))
+cosh = define_unary_op("cosh", lambda x: math.cosh(x))
```

Comment:
Do we really need to support every single one of these math ops to get lambdify to work? I don't see any use case where someone is calling these inverse or hyperbolic trig functions on a time-series.

Reply:
I would consider also defining all these math ops in a separate math.py file (and under a csp.math import path). It's bloating the code in baselib and they are too specialized for the average use case, much like the stats functions which reside in their own file (csp.stats).

Reply:
we can move all the math operations to csp.math, i think thats a good idea. Including `abs`, `exp`, and `log`.

Reply:
done, isnt really a breaking change as all are exported to the top level (e.g. `csp.and_`)

Reply:
Noticed that you still have the cpp impl in baselibimpl.cpp and cspbaselibimpl.cpp, could you separate these to mathimpl.cpp and cspmathimpl.cpp?